File: src/main/java/net/projecteuler/barreiro/problem/Solver038.java
/* COPYRIGHT (C) 2014 barreiro. All Rights Reserved. */

package net.projecteuler.barreiro.problem;

import net.projecteuler.barreiro.algorithm.util.LongUtils;

import java.util.Comparator;
import java.util.function.LongPredicate;

import static java.util.Arrays.stream;
import static java.util.stream.LongStream.rangeClosed;
import static net.projecteuler.barreiro.algorithm.Combinatorics.partitionStream;
import static net.projecteuler.barreiro.algorithm.Combinatorics.permutationStream;
import static net.projecteuler.barreiro.algorithm.util.LongUtils.intSqrt;
import static net.projecteuler.barreiro.algorithm.util.StreamUtils.longSet;

/**
 * Take the number 192 and multiply it by each of 1, 2, and 3:
 * 192 × 1 = 192
 * 192 × 2 = 384
 * 192 × 3 = 576
 * By concatenating each product we get the 1 to 9 pandigital, 192384576. We will call 192384576 the concatenated product of 192 and (1,2,3)
 * <p>
 * The same can be achieved by starting with 9 and multiplying by 1, 2, 3, 4, and 5, giving the pandigital, 918273645, which is the concatenated product of 9 and (1,2,3,4,5).
 * <p>
 * What is the largest 1 to 9 pandigital 9-digit number that can be formed as the concatenated product of an integer with (1,2, ... , n) where n > 1?
 *
 * @author barreiro
 */
public class Solver038 extends ProjectEulerSolver {

    private static final Comparator<long[]> REVERSE_COMPARATOR = (l1, l2) -> {
        for ( int i = l1.length - 1; i >= 0; i-- ) {
            if ( l1[i] != l2[i] ) {
                return (int) ( l2[i] - l1[i] );
            }
        }
        return 0;
    };

    public Solver038() {
        this( 9 );
    }

    public Solver038(long n) {
        super( n );
    }

    // --- //

    private boolean hasSameFactor(long... partition) {
        LongPredicate reverseNatural = p -> rangeClosed( 1, partition.length ).allMatch( n -> partition[(int) n - 1] / p == partition.length - n + 1 );
        return stream( partition ).filter( l -> stream( partition ).allMatch( p -> l != 0 && p % l == 0 && p / l < N ) ).anyMatch( reverseNatural );
    }

    private boolean solveSingle(long partitions, long... permutation) {
        return partitionStream( permutation, (int) partitions ).filter( l -> stream( l ).allMatch( a -> a > 10 ) ).anyMatch( this::hasSameFactor );
    }

    // different approach: generate pandigital numbers, order them and then try to find a multiplier from one of the possible partitions
    public long solve() {
        return permutationStream( longSet( rangeClosed( 1, N ) ) ).sorted( REVERSE_COMPARATOR ).filter( p -> rangeClosed( 2, intSqrt( N ) ).anyMatch( n -> solveSingle( n, p ) ) ).mapToLong( LongUtils::fromDigits ).findFirst().orElse( 0L );
    }
}

File: src/main/rust/euler/solver038.rs
// COPYRIGHT (C) 2017 barreiro. All Rights Reserved.
// Rust solvers for Project Euler problems

use algorithm::cast::Cast;
use algorithm::digits::{concatenation, Digits};
use algorithm::root::{ceil_sqrt_u64, int_log_10, pow_10};
use Solver;

/// Take the number `192` and multiply it by each of `1, 2, and 3`:
/// ```
/// 192 × 1 = 192
/// 192 × 2 = 384
/// 192 × 3 = 576
/// ```
/// By concatenating each product we get the `1` to `9` pandigital, `192384576`. We will call `192384576` the concatenated product of `192` and `(1,2,3)`
///
/// The same can be achieved by starting with `9` and multiplying by `1, 2, 3, 4, and 5`, giving the pandigital, `918273645`, which is the concatenated product of `9` and `(1,2,3,4,5)`.
///
/// What is the largest `1` to `9` pandigital `9-digit` number that can be formed as the concatenated product of an integer with `(1,2, ... , n)` where `n > 1`?
pub struct Solver038 {
    pub n: u64,
}

impl Default for Solver038 {
    fn default() -> Self {
        Self { n: 9 }
    }
}

impl Solver for Solver038 {
    fn problem_name(&self) -> &str { "Pandigital multiples" }

    fn solve(&self) -> i64 {
        let natural_product_concatenation = |value, size| {
            let (mut n, mut result) = (value, value);
            loop {
                if int_log_10(result) >= size {
                    return Digits::from(result);
                }
                n += value;
                result = concatenation(result, n);
            }
        };

        // generate the natural product concatenation and check if it's a pandigital. find the first according to the most significant digit ordering
        decrementing_by_most_significant(self.n, ceil_sqrt_u64(self.n)).map(|v| natural_product_concatenation(v, self.n)).filter(|d| d.len() == self.n.as_usize()).find(Digits::is_pandigital).as_i64()
    }
}

// --- //

/// creates an iterator that returns elements ordered by most significant digit: 999 ... 900 -> 99 ... 90 -> 9 -> 899 ... 800 -> 89 .. 80 -> 8 -> 799
fn decrementing_by_most_significant(value: u64, scale: u64) -> impl Iterator<Item=u64> {
    DecrementingByMostSignificant { range: Box::new(0..0), size: 0, value: value + 1, scale }
}

struct DecrementingByMostSignificant {
    range: Box<dyn Iterator<Item=u64>>,
    size: u64,
    value: u64,
    scale: u64,
}

impl Iterator for DecrementingByMostSignificant {
    type Item = u64;

    fn next(&mut self) -> Option<u64> {
        self.range.next().or_else(|| {
            if self.size > 0 {
                self.size -= 1;
            } else {
                self.size = self.scale;
                self.value -= 1;
            }
            if self.value > 0 {
                self.range = Box::new((self.value * pow_10(self.size)..(self.value + 1) * pow_10(self.size)).rev());
                self.range.next().filter(|&v| v > 1)
            } else {
                None
            }
        })
    }
}
